[Build][Runtime] Add a new threading library.

al45tair opened the pull request:
A fair bit of this is really shuffling things around — that's largely true for the `swift::Mutex`, which was an existing thing. `swift::Thread` is really a placeholder for when we need more support — right now, we don't need any of the functionality that e.g. `std::thread` has, but we _do_ need some functionality it doesn't have. It's also desirable to be able to choose which threading package to use, especially for embedded use or for testing purposes, and if we make everything defer to the C++ library we'll only be able to use whatever the C++ library is using.

The main new things are:
- You can ask to use a specific threading package.
- The threading code is in its own library, so no more `#include`s of implementation files in order to bring them in to the standard library or concurrency code.
- `swift::Lazy` now uses `swift::once` rather than having its own implementation that uses a macro.
- New `swift::once` implementations for Linux and Windows that should be somewhat faster on those platforms.
- Improved error handling, so the threading code raising errors can hook into the error handling used in the places it's being used.
- Adjusted the TLS support code slightly, to avoid trying to put TLS keys in template arguments, as that isn't guaranteed to work.

al45tair committed:
[Build][Runtime] Replace SWIFT_STDLIB_SINGLE_THREADED_RUNTIME.

SWIFT_STDLIB_SINGLE_THREADED_RUNTIME is too much of a blunt instrument here.
It covers both the Concurrency runtime and the rest of the runtime, but we'd
like to be able to have e.g. a single-threaded Concurrency runtime while
the rest of the runtime is still thread safe (for instance).

So: rename it to SWIFT_STDLIB_SINGLE_THREADED_CONCURRENCY and make it just
control the Concurrency runtime, then add a SWIFT_STDLIB_THREADING_PACKAGE
setting at the CMake/build-script level, which defines
SWIFT_STDLIB_THREADING_xxx where xxx depends on the chosen threading package.

This is especially useful on systems where there may be a choice of threading
package that you could use.

rdar://90776105

al45tair committed:
[Runtime] Remove all use of read/write locks.

Read/write locks are not as good as you'd think; a simple mutex is better in
almost all cases.

rdar://90776105

al45tair committed:
[Threading] Create new threading library and use it.

Moved all the threading code to one place.  Added explicit support for
Darwin, Linux, Pthreads, C11 threads and Win32 threads, including new
implementations of Once for Linux, Pthreads, C11 and Win32.

rdar://90776105

al45tair commented:
@swift-ci Please test

al45tair commented:
@swift-ci Please test

al45tair commented:
@swift-ci Please test

al45tair committed:
[Unittests] Make the Threading unit tests work.

The threading unit tests currently just check the operation of Mutex.
This used to be part of the runtime tests, but now it's a separate
library we can test it separately.

rdar://90776105

al45tair committed:
[Unittests] Add more threading library tests.

Added explicit tests for `swift::once` and also for the ulock implementation
used on Linux.

rdar://90776105

al45tair commented:
@swift-ci Please test

al45tair committed:
[Threading] Fix case sensitivity bug.

The file is called Nothreads.cpp, not NoThreads.cpp.

rdar://90776105

al45tair committed:
[Threading] Auto-detect threading in the preprocessor.

Some things (LLDB's language plugin, for instance), include Swift headers
from outside of our build system, and may not know to define one of the
threading package defines.

rdar://90776105

al45tair commented:
@swift-ci Please test

al45tair commented:
@swift-ci Please test windows platform

al45tair committed:
[Threading] A couple of header file fixes.

Some declarations were missing from non-Darwin sources, and the Linux
header was slightly broken.

rdar://90776105

al45tair committed:
[Threading] Fix the Linux build.

A few fixes specifically for the Linux build.

rdar://90776105

al45tair commented:
@swift-ci Please test

al45tair commented:
@swift-ci Please test

al45tair committed:
[Build] Move initialization of SWIFT_HOST_VARIANT_SDK further up.

We want this initialized when we decide which threading package to
use.  This isn't a problem on Darwin or Linux where we're using the
build script, but on Windows it relies on the defaulting mechanism,
which didn't happen until further down the `CMakeLists.txt`.

rdar://90776105

al45tair committed:
[Build] Fix static linking.

When we static link against libswiftCore, we need it to contain the
actual object modules from the threading library (when it's dynamically
linked, the shared object will already contain those modules because it's
been linked against libswiftThreading), which means that the threading
library in the stdlib needs to be an OBJECT_LIBRARY rather than a STATIC
library (otherwise people will have to statically link against the
threading library as well, and that'll need to be installed, and we don't
want that).

rdar://90776105

al45tair committed:
[Build] Fix the name of the threading package setting.

I'd renamed the setting, but failed to update build-presets.ini.

rdar://90776105

al45tair commented:
@swift-ci Please test

al45tair commented:
@swift-ci Please test Windows platform

al45tair committed:
[Threading][Windows] Fix missing close brace.

Apparently I missed a close brace from the end of Win32.h.

rdar://90776105

al45tair commented:
@swift-ci Please test Windows platform

al45tair commented:
@swift-ci Please test Windows platform

al45tair committed:
[Threading][Windows] Fix another Windows niggle.

Looks like I forgot to update the Win32.h implementation.

rdar://90776105

al45tair commented:
@swift-ci Please test Windows platform

al45tair commented:
@swift-ci Please test Windows platform

al45tair committed:
[Threading][Windows] More header file fixes.

rdar://90776105

al45tair commented:
@swift-ci Please test Windows platform

al45tair commented:
@swift-ci Please test Windows platform

al45tair commented:
@swift-ci Please test Windows platform

al45tair commented:
@swift-ci Please test Windows platform

al45tair commented:
@swift-ci Please test Windows platform

al45tair committed:
[Threading][Windows] Undefine Yield and ERROR.

`windows.h` defines a couple of very unhelpfully named macros.
Undefine them after including it.

rdar://90776105

al45tair commented:
@swift-ci Please test Windows platform

al45tair commented:
@swift-ci Please test Windows platform

al45tair committed:
[Runtime][Windows] Need to link with Synchronization.lib.

We're using some functions from Synchronization.lib now, after the
threading changes.

rdar://90776105

al45tair commented:
@swift-ci Please test Windows platform

al45tair committed:
[Threading][Windows] Don't include <windows.h>

We can't safely include <windows.h> because it defines a large number
of macros, some of which clash with things in the Swift source tree,
and others of which clash with things in the LLVM source tree.  Sadly
we *also* can't just include the Windows headers we need, because they
pull in some of the problematic macros.

In this instance, the best thing seems to be to grab the definitions
for the types and functions we are going to use and put them in their
own header file.  If we define them correctly, then #including
<windows.h> before or after this header won't have any adverse effects.

rdar://90776105

al45tair commented:
@swift-ci Please test Windows platform

al45tair commented:
@swift-ci Please test Windows platform

al45tair commented:
@swift-ci Please test Windows platform

al45tair committed:
[Runtime][Windows] A couple of files need to include <windows.h>

We shouldn't include <windows.h> implicitly from .cpp files, but should
do it directly so that we know it's there.

Also, if we're including <windows.h>, do it at the top of the file.

rdar://90776105

al45tair commented:
@swift-ci Please test Windows platform

al45tair commented:
@swift-ci Please test Windows platform

al45tair commented:
@swift-ci Please test Windows platform

al45tair commented:
@swift-ci Please test Windows platform

al45tair committed:
[Concurrency][Threading] Remove use of platform thread functions.

The concurrency library can use the new threading library functions,
which avoids the problem of including <windows.h>.

rdar://90776105

al45tair commented:
@swift-ci Please test

al45tair committed:
[Threading][Windows] Try to avoid declaring _RTL_SRWLOCK

Declaring _RTL_SRWLOCK ourselves causes clashes with <windows.h>.
Rather than doing that, declare an equivalent struct, and some overloads.

rdar://90776105

al45tair commented:
@swift-ci Please test Windows platform

al45tair committed:
[Concurrency][Windows] Add a couple of includes for Windows.

Actor.cpp does need <io.h> still, and Task.cpp should have been including
<windows.h>

rdar://90776105

al45tair commented:
@swift-ci Please test Windows platform

al45tair committed:
[Concurrency][Windows] Remove unnecessary includes.

TaskLocal.cpp doesn't need <handleapi.h> or <processthreadsapi.h>, both of
which want <windows.h>, which isn't included any more.

rdar://90776105

al45tair commented:
@swift-ci Please test Windows platform

al45tair committed:
[Threading][Windows] Link the unit test against Synchronization.lib

We need to link the threading unit test against Synchronization.lib
to pick up the Win32 APIs we're using.

rdar://90776105

al45tair commented:
@swift-ci Please test Windows platform

al45tair committed:
[UnitTests][Windows] Link with Synchronization.lib

The runtime unit tests also need to link with Synchronization.lib.

rdar://90776105

al45tair commented:
@swift-ci Please test Windows platform

al45tair commented:
@swift-ci Please test Windows platform

al45tair commented:
@swift-ci Please test

al45tair commented:
@swift-ci Please benchmark

al45tair committed:
[Threading] Remove thread_get_main().

Removing thread_get_main() means we don't need a static initializer on
Darwin, which means we can delete Darwin.cpp as well.  We can also delete
Nothreads.cpp while we're about it, because there's nothing in that file.

rdar://90776105

al45tair committed:
[Threading] Put everything through git clang-format.

Just formatting changes.

rdar://90776105

al45tair committed:
[Threading] Fix some problems with the C11 threading code.

These changes are needed to get things building with a C11 threads shim
header on macOS.

rdar://90776105

al45tair commented:
@swift-ci Please test

al45tair commented:
@swift-ci Please benchmark

al45tair commented:
@swift-ci Please test

al45tair committed:
[Threading] Don't use TLS keys as template arguments.

There's no guarantee that e.g. pthread_key_t is an integral type.  It could
be some kind of struct, or some other thing that isn't valid as a template
argument.

rdar://90776105

al45tair commented:
@swift-ci Please test

al45tair commented:
@swift-ci Please test

al45tair committed:
[Threading] Move stack bounds fetching into the threading library.

Fetching the current stack bounds is done using threading functions, so
should be in the threading library.

rdar://90776105

al45tair committed:
[Concurrency] Remove redundant include.

Clean up an unnecesary include.

rdar://90776105

al45tair commented:
@swift-ci Please test

al45tair committed:
[Threading] A few include tweaks.

The threading implementations need to include Impl.h, not just their
specific variant, to pick up the `stack_bounds` type.

Also fix a newly exposed problem with Atomic.h.

rdar://90776105

al45tair commented:
@swift-ci Please test

grynspan commented:
I haven't taken a deep dive yet, but I'm wondering about the necessity of the platform-specific implementations. Since Linux and Windows both have C++'s `std::thread` (and `std::mutex` and and and…), would it make sense to just… use that for every non-Apple platform? I can buy that Darwin may need to be specialized.

Given Swift's lack of official support for platforms other than Darwin and Linux, it seems to me like we should lean on the STL as much as possible rather than rolling our own equivalent implementations.

grynspan commented on include/swift/Runtime/EnvironmentVariables.h:
```diff
@@ -14,25 +14,25 @@
-    return name ## _variable;                           \
+#define VARIABLE(name, type, defaultValue, help)                               \
+  inline type name() {                                                         \
+    swift::once(initializeToken, initialize, nullptr);                         \
```
❤️

grynspan commented on include/swift/Threading/Once.h:
```diff
@@ -0,0 +1,36 @@
+/// Runs the given function with the given context argument exactly once.
+/// The predicate argument must refer to a global or static variable of static
+/// extent of type swift::once_t.
+inline void once(once_t &predicate, void (*fn)(void *),
```
May I suggest adding a templatized overload that takes an arbitrary functor? See for instance what I did with `_swift_withWin32DbgHelpLibrary()`. An overload means we can just directly pass a lambda with captures rather than worrying about `context`.)

al45tair replied:
That seems like a good idea.

grynspan commented on stdlib/public/stubs/Stubs.cpp:
```diff
@@ -537,66 +535,10 @@ __swift_bool swift_stdlib_isStackAllocationSafe(__swift_size_t byteCount,
-#elif defined(__OpenBSD__)
-  stack_t sinfo;
-  if (pthread_stackseg_np(pthread_self(), &sinfo) != 0) {
+  swift::Thread::StackBounds bounds = swift::Thread::stackBounds();
```
Nitpick, maybe have `stackBounds()` return an `llvm::Optional` or `std::optional`? I personally dislike using integer `0` as a magic value.

grynspan replied:
Gracias. It's a nitpick to be sure—you're right that zero is not likely to be the stack's lower bound. :)

al45tair commented:
I've updated the comment at the top with some notes on the changes that are hopefully useful for reviewers.

gottesmm commented on include/swift/Runtime/ThreadLocal.h:
```diff
@@ -20,14 +20,15 @@
 #include "ThreadLocalStorage.h"
+#include "Once.h"
 
 /// SWIFT_RUNTIME_SUPPORTS_THREAD_LOCAL - Does the current configuration
```
Can you update the variable name in the comment?

al45tair replied:
Pretty sure that was already fixed.

al45tair committed:
[Threading] Add the ability to use any C++ callable in swift::once().

This allows the use of C++ lambdas and functors, in addition to plain
old functions.

rdar://90776105

al45tair committed:
[Threading] Fix a few things following rebase.

After rebasing, a few things broke.  Fix them.

rdar://90776105

al45tair committed:
[Threading] Use llvm::Optional<> rather than making a zero lower bound special.

While most systems aren't going to have their stack bottom at zero,
using llvm::Optional<> here is cleaner.

rdar://90776105

al45tair commented:
@swift-ci Please test

gottesmm commented:
@al45tair can you make sure the standalone minimal test still passes? I think it uses the single threaded flag.

al45tair commented:
preset=stdlib_S_standalone_minimal_macho_x86_64,build,test
@swift-ci please test with toolchain and preset

al45tair commented:
Hmmm. "Can't find CMake" seems like not my fault. I'll give it another go, then go look what the problem might be.

al45tair commented:
preset=stdlib_S_standalone_minimal_macho_x86_64,build,test
@swift-ci please test with toolchain and preset

kubamracek commented:
LGTM (sorry, it took a while to go over the large diff).

Given that the PR testing for "stdlib minimal" seems to have some infrastructure issues, I am going to run the "preset testing" locally to ensure we're not breaking it, I'll report back.

al45tair committed:
[Build] Fix some issues with the standalone library build.

When we're building static libraries, we don't need to include the
threading objects in both Concurrency and Core, and we also don't need
two copies of the threading library's fatal error handler.

rdar://90776105

al45tair commented:
@swift-ci Please test

al45tair commented:
preset=stdlib_S_standalone_minimal_macho_x86_64,build,test
@swift-ci please test with toolchain and preset

al45tair commented:
This passed local testing of the minimal preset just now, FWIW:

```
Testing Time: 456.15s
  Excluded         : 14471
  Unsupported      :   230
  Passed           :   289
  Expectedly Failed:     1

49 warning(s) in tests
-- check-swift-validation-only_executable-freestanding-x86_64 finished --
--- Finished tests for swift ---
```

al45tair commented:
Looks like preset testing in CI is still broken.

gottesmm commented:
@shahmishal can you prioritize fixing preset testing in CI? We need it to work to test PRs like this?

gottesmm commented:
If the preset testing worked locally... I am happy.

al45tair commented:
@swiftci please test linux platform

al45tair commented:
@swift-ci Please test Linux platform

al45tair commented:
@swift-ci Please test

al45tair committed:
[Build][Concurrency] Fix Concurrency build to work for Linux.

`SWIFT_BUILD_STATIC_STDLIB` is not mutually exclusive with
`SWIFT_BUILD_DYNAMIC_STDLIB`, and Linux sets both, so we can't use
`SWIFT_BUILD_STATIC_STDLIB` to conditionalise things.

The linker error about duplicate definitions for the threading error
handler was happening because we were forced to include the object
containing that symbol; moving it to another object should fix that.

And it turns out there's a way to get CMake to include the threading
object library only when building a shared object, so use that too.

rdar://90776105

al45tair commented:
The Linux issue has to do with the fact that I'd not appreciated that `SWIFT_BUILD_STATIC_STDLIB` and `SWIFT_BUILD_DYNAMIC_STDLIB` weren't mutually exclusive. Hopefully the latest change fixes that.

Local testing of the preset passes still:
```
Testing Time: 452.37s
  Excluded         : 14471
  Unsupported      :   230
  Passed           :   289
  Expectedly Failed:     1

49 warning(s) in tests
-- check-swift-validation-only_executable-freestanding-x86_64 finished --
--- Finished tests for swift ---
```

kateinoigakukun commented:
Thank you @al45tair for your great work!! The `none` thread variant is really helpful for us Wasm support!

al45tair commented:
@kateinoigakukun Glad to know it's helpful. It's been reverted for now because of a few CI failures (the Linux problem you sent a PR for, plus some simulator related problems). I'll get it re-submitted ASAP.
